### ISSUE TITLE
Training batch sampling not reproducible across Python patch versions

### ISSUE THREAD

**Issue by rdrip** [howervoo/speechturn#16]
## Context

The training loop in `training.py` uses Python's `random.randrange` to
sample batches:

```python
batches = [
    collate_examples(
        [examples[random.randrange(len(examples))] for _ in range(options.batch_size)], cfg
    )
    for _ in range(options.gradient_accumulation)
]
```

`seed_everything` seeds `random.seed(seed)` at the start, and
`load_checkpoint(restore_rng=True)` restores the Python `random` state
during resume. The `test_resume_matches_uninterrupted_training` test verifies
exact loss and weight parity between continuous and interrupted runs.

## Problem

This works within a single Python installation, but `random.randrange` is
not guaranteed to produce the same sequence across Python patch versions. The
CPython `random` module's internal `_randbelow` method changed between
3.11.x and 3.12.x (the rejection sampling threshold was adjusted).

In practice: training for 100 steps on Python 3.11.9, saving a checkpoint at
step 50, then resuming on Python 3.11.10 produces a different batch sequence
than training continuously on 3.11.10 from step 1. The RNG state (Mersenne
Twister output) is identical, but `randrange` consumes a different number
of random bits per call depending on the implementation.

## Steps to reproduce

```bash
# On Python 3.11.9:
speechturn train data/manifest.jsonl model-50.pt --steps 50 --seed 42
speechturn train data/manifest.jsonl model-100.pt --steps 100 --seed 42 --resume model-50.pt

# On Python 3.12.3 (same data, same torch, same numpy):
speechturn train data/manifest.jsonl model-100-ref.pt --steps 100 --seed 42

# Compare: model-100.pt and model-100-ref.pt differ
```

## Why this matters

The README states: "To reproduce an interrupted run, keep the data, batch size,
accumulation, optimizer settings and software environment unchanged." The
software environment clause implicitly covers the Python patch version, but
this is not documented and easy to overlook when upgrading a container image.

## Suggested fix

Use `numpy.random.Generator` (or `torch.randint`) for batch index sampling
instead of `random.randrange`. Both NumPy and PyTorch guarantee stable RNG
output across patch versions for a given major.minor release. The training
loop already restores NumPy and PyTorch RNG states via `seed_everything`,
so switching the sampling source would make cross-version resume work without
changing the user-facing API.
